src/lib/database/official/syncTricks.ts

```diff
@@ -0,0 +1,193 @@
+  // What we *can* do however is PUT new Tricks into the Table that have trickStatus set to archived and then delete the original trick.
+
+  // First, make sure all references to this trick are changed to the archived counterpart.
+  // Currently, external references can be found in:
+  // - metadata
+  // - recommendedPrequisites
+  // - variationOf
+  // - tricks belonging to a combo
+  
+  // Immediately invoked promises.
+  // TODO: Extract into separate functions
+  // One might consider running these in parallel, but there is the risk of a 
+  // race condition / working on something that is currently being modified.
+  // To keep things safe it is probably better to run those in sequence.
+
+  const handleMetadata = async() => {
+    const metadataResult = await db.metadata.get([trickId, from, "Trick"])
+    if(!metadataResult) {
+      return
+    }
+    // Metadata exists.
+    // Compose new Metadata
+    const newMetadata = {
+      ...metadataResult,
+      trickStatus: to
+    }
+    await db.metadata.put(newMetadata)
+    // Then delete the old metadata
+    await db.metadata.delete([trickId, from, "Trick"])
+  }
+
+  const handleRecommendedPrerequisites = async () => {
+    await db.tricks.where("recommendedPrerequisites").equals([trickId, from]).modify( e => {
+      if(!e.recommendedPrerequisites) {
+        return
+      }
+
+      e.recommendedPrerequisites = e.recommendedPrerequisites.map( ([id, status]) => {
+        if(id === trickId && status === from) {
+          return [id, to]
+        }
+        return [id, status]
+      } )
+      
+    })
+  }
+
+  const handleVariationOf = async () => {
+    await db.tricks.where("variationOf").equals([trickId, from]).modify( e => {
+      if(!e.variationOf) {
+        return
+      }
+
+      e.variationOf = e.variationOf.map( ([id, status]) => {
+        if(id === trickId && status === from) {
+          return [id, to]
+        }
+        return [id, status]
+      })
+    })
+  }
+
+  const handleTrickReferences = async () => {
+    await db.combos.where("tricks").equals([trickId, from]).modify( e => {
+      e.tricks.map( ([id, status]) => {
+        if(id === trickId && status === from) {
+          return [id, to]
+        }
+        return [id, status]
+      })
+    })
+  }
+
+
+  try {
+    await handleMetadata()
+  }catch(err) {
+    console.error(`[Official Sync] Failed to update metadata for ${trickId}`, err)
+    
+  }
+  try {
+    await handleRecommendedPrerequisites()
+  }catch(err) {
+    console.error(`[Official Sync] Failed to update recommended Prerequisites referencing ${trickId}`, err)
+  }
+  try {
+    await handleVariationOf()
+  }catch(err) {
+    console.error(`[Official Sync] Failed to update variationOf referencing ${trickId}`, err)
+  }
+  try {
+    await handleTrickReferences()
+  }catch(err) {
+    console.error(`[Official Sync] Failed to combos referencing trick ${trickId}`, err)
+  }
```

Comment:
To me this looks like all the workings of the top level function (`moveTrick`) should be one ACID-Transaction (Atomic, Consistent, Independent, Durable) on the database. The big two ones here are Atomic and Consistent. Both are violated, as it is possible to run into errors with some of the individual database transactions but still execute the others (not atomic, atomic means either all or nothing) which could potentially result in inconsistent database entries.

The solution could be to just try and do all the transactions with the database but only commit the changes one all of them were successful.

But then also I am seeing this function for the first time, removed from most of its context, so I might be completely off in my assessment.

Reply:
The commiting of the canges could, if my understanding of Dexie is correct, even be left out of this function completely and be left to the caller of the function, allowing the caller to create larger ACID-transactions.

Reply:
Would being Atomic in this case be *really* desirable? 

Instead of having e.g. one Trick's Metadata fail to get moved to the "archived" status you would have a situation where all Tricks would not get synced.

Reply:
Yes! At least parts. I would argue that the whole thing should be one large atomic transaction, which either updates everything properly or does nothing at all and raises an error which gets reported to either the user or, better, us. This is debatable though.

What I think needs to change for sure is that at least parts of the function should become atomic transactions. The main one I just found is the `handleMetadata` function, which makes a `get`, a `put` and a `delete` interaction with the database. All these interactions are directly linked to each other and they should be one atomic transaction so that the database is still in a consistent state after the function finishes. This is because the data in the database could change between any of the steps or an error could occur while deleting, leaving the Metadata-Schema with a bunch of "duplicate" entries that should not exist.